
--- numlib/numsup.h	2015-10-26 05:17:33.000000000 +0100
+++ numlib/numsup.h	2016-10-04 10:10:28.732000000 +0200
@@ -324,7 +324,7 @@
 #define error_program g_log->tag
 extern plan10 set_exe_path(char *arg0);
 
-extern plan10 ATTRIBUTE_NORETURN error(char *fmt, ...);
+extern plan10 error(char *fmt, ...) __attribute__((noreturn));
 extern plan10 warning(char *fmt, ...);
 extern plan10 verbose(int level, char *fmt, ...);
 
