diff --git lib/package_script.c lib/package_script.c
index a8c05dcf..919f821b 100644
--- lib/package_script.c
+++ lib/package_script.c
@@ -40,9 +40,10 @@ pkginst_pkg_exec_buffer(struct pkginst_handle *xhp,
                     bool update)
 {
        ssize_t ret;
-       const char *tmpdir, *version;
+       const char *tmpdir, *version, *shell;
        char *pkgname, *fpath;
        int fd, rv;
+       struct stat st;
 
        assert(blob);
        assert(pkgver);
@@ -100,7 +101,12 @@ pkginst_pkg_exec_buffer(struct pkginst_handle *xhp,
        version = pkginst_pkg_version(pkgver);
        assert(version);
 
-       rv = pkginst_file_exec(xhp, "/bin/sh", fpath, action, pkgname, version,
+       // Fallback
+       shell = "/bin/sh";
+       if (stat(shell, &st) < 0) {
+               shell = "/bin/dash";
+       }
+       rv = pkginst_file_exec(xhp, shell, fpath, action, pkgname, version,
                            update ? "yes" : "no",
                            "no", xhp->native_arch, NULL);
        free(pkgname);
